File: BinaryTrees/longest_zig_zag.py
from typing import Optional

from tree_node import TreeNode


class LongestZigZag:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        def dfs(node: Optional[TreeNode], left_length: int, right_length: int) -> int:
            nonlocal max_length
            if not node:
                return

            max_length = max(max_length, left_length, right_length)
            dfs(node.right, right_length + 1, 0)
            dfs(node.left, 0, left_length + 1)

        max_length = 0
        dfs(root, 0, 0)
        return max_length

    # Time complexity: O(n) where n is the number of nodes in the tree,
    # since we visit each node exactly once.
    # Space complexity: O(h) where h is the height of the tree,
    # due to the recursion stack. 
    # In the worst case of a skewed tree, this can be O(n).

File: BinaryTrees/LongestZigZag.java
package BinaryTrees;

public class LongestZigZag {
  private int maxLength;

  public int longestZigZag(TreeNode root) {
    maxLength = 0;
    dfs(root, 0, 0);
    return maxLength;
  }

  private void dfs(TreeNode node, int left, int right) {
    if (node == null) {
      return;
    }

    maxLength = Math.max(Math.max(left, right), maxLength);
    dfs(node.left, right + 1, 0);
    dfs(node.right, 0, left + 1);
  }
}

// Time Complexity: O(n) where n is the number of nodes in the tree
// Space Complexity: O(h) where h is the height of the tree, due to the
// recursive call stack. In the worst case of a skewed tree, this can be O(n).